### PR TITLE
[raysgd] Custom training operator

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Provide an abstraction for stateful custom training loops. See requirement doc here:

https://docs.google.com/document/d/1Cjp8Xtu9dG_vOO9yyrEIGL2MypgSsJ5wApzBa_BkUOQ/edit#

Need to:
- [x] write documentation
- [x] write tests

This PR leaves a couple usability stories to do:

1. `train_epoch(num_steps)` should let you intercept training during an epoch for custom methods such as checkpoints. Do we continue training from the point we left off?
2. If my 100 batches returns some loss, how do I aggregate this? 
3. In the current interface, custom learning rate schedulers are unwieldy. Is there a clear way of handling them?

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Review by maximsmol — CHANGES_REQUESTED** [review]
Suggesting some documentation clarification.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 1]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
 
-.. code-block:: python
+Custom Training and Validation (Operators)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    trainer.train()
+``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexibility similar to using PyTorch natively.
+This is done via the :ref:`ref-pytorch-operator` interface.
+
+For both training and validation, there are two granularities that you can provide customization - per epoch or per batch. These correspond to ``train_batch``,
```
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/raysgd/raysgd_pytorch.rst]: @maximsmol FYI you can click on the line number to suggest changes inline

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 2]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
+            Returns:
+                A dict of metrics. Defaults to "loss" and "num_samples",
+                    corresponding to the total number of datapoints in the batch.
+            """
+            real_label = 1
+            fake_label = 0
+            discriminator, generator = self.models
+            optimD, optimG = self.optimizers
+
+            discriminator.zero_grad()
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: comments with high-level descriptions of what's happening in each of these blocks would be helpful

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 3]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
+        stats = trainer.train()
+        print(stats)
+
+See the `DCGAN example <https://github.com/ray-project/ray/blob/master/python/ray/util/sgd/pytorch/examples/dcgan.py>`__ for an end to end example. It constructs two models and two optimizers and uses a custom training operator to provide a non-standard training loop.
+
+
+Training/Validation
+-------------------
+
+Now that the trainer is constructed, you'll naturally want to train the model.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Now that the trainer is constructed, here's how to train the model.
```
"naturally" sounds like fluff, always good to use simple language where possible for things like these IMO

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 4]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
+.. code-block:: python
 
-Shutting down training
-----------------------
+    for i in range(10):
+        metrics = trainer.train()
+        val_metrics = trainer.validate()
+
+
+Each ``train`` takes one pass over the training data, and each ``validate`` call runs the model on the validation data passed in by the ``data_creator``. Provide a custom training operator (:ref:`raysgd-custom-training`) to calculate custom metrics or customize the training/validation process.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Each ``train`` call makes one pass over the training data and each ``validate`` call runs the model on the validation data passed in by the ``data_creator``. Provide a custom training operator (:ref:`raysgd-custom-training`) to calculate custom metrics or customize the training/validation process.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 5]
```diff
@@ -242,22 +334,21 @@ To specify particular parameters for ``amp.initialize``, you can use the ``apex_
...
 
-Note that if using a custom training function, you will need to manage loss scaling manually.
+Note that if using a custom training operator (:ref:`raysgd-custom-training`), you will need to manage loss scaling manually.
 
 
 Distributed Multi-node Training
 -------------------------------
 
-You can scale out your training onto multiple nodes without making any modifications to your training code. To train across a cluster, simply make sure that the Ray cluster is started.
+You can scale out your training onto multiple nodes without making any modifications to your training code. To train across a cluster, simply make sure that the Ray cluster is started. You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ or `manually <using-ray-on-a-cluster.html>`_.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
You can scale your training to multiple nodes without making any modifications to the training code. To train across a cluster, simply make sure that the Ray cluster is started. You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ or `manually <using-ray-on-a-cluster.html>`_.
```

Comment (edoakes) [reply doc/source/raysgd/raysgd_pytorch.rst]: This was a bit confusing to me because you say "simply make sure that the Ray cluster is started" then end the sentence without indicating that you have to change the code to connect to it. I thought that was going to magically work - readers might stop there and think that too

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 6]
```diff
@@ -242,22 +334,21 @@ To specify particular parameters for ``amp.initialize``, you can use the ``apex_
...
 
-You can start a Ray cluster `via the Ray cluster launcher <autoscaling.html>`_ or `manually <using-ray-on-a-cluster.html>`_.
+In your program, you'll need to connect to this cluster via ``ray.init``:
 
-.. code-block:: bash
+.. code-block:: python
 
-    ray up CLUSTER.yaml
-    ray submit train.py --args="--address='auto'"
+    ray.init(address="auto")  # or a specific redis address.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]: ... "of the form <ip-address:port>"

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 7]
```diff
@@ -310,22 +404,37 @@ Advanced: Hyperparameter Tuning
 Simultaneous Multi-model Training
 ---------------------------------
 
-In certain scenarios such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator``, ``optimizer_creator``, and ``scheduler_creator`` to return multiple values.
-
-If multiple models, optimizers, or schedulers are returned, you will need to provide a custom training function (and custom validation function if you plan to call ``validate``).
+In certain scenarios such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator``, ``optimizer_creator``, and ``scheduler_creator`` to return multiple values. Provide a custom TrainingOperator (:ref:`raysgd-custom-training`) to train across multiple models.
```
Comment (edoakes) [inline doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
In certain scenarios, such as training GANs, you may want to use multiple models in the training loop. You can do this in the ``PyTorchTrainer`` by allowing the ``model_creator``, ``optimizer_creator``, and ``scheduler_creator`` to return multiple values. Provide a custom TrainingOperator (:ref:`raysgd-custom-training`) to train across multiple models.
```

File: python/ray/util/sgd/pytorch/examples/dcgan.py [hunk 8]
```diff
@@ -114,119 +95,137 @@ def forward(self, x):
...
+
+        preds = np.zeros((N, 10))
+        for i, batch in enumerate(dataloader, 0):
+            batch = batch.type(torch.FloatTensor)
+            batchv = Variable(batch)
+            batch_size_i = batch.size()[0]
+            preds[i * batch_size:i * batch_size +
+                  batch_size_i] = get_pred(batchv)
+
+        # Now compute the mean kl-div
```
Comment (edoakes) [inline python/ray/util/sgd/pytorch/examples/dcgan.py]: This looks weird without any preceding comments. Maybe add some more similar ones for other blocks?

File: python/ray/util/sgd/pytorch/examples/dcgan.py [hunk 9]
```diff
@@ -240,7 +239,7 @@ def train_example(num_replicas=1, use_gpu=False, test_mode=False):
         "--address",
         required=False,
         type=str,
-        help="the address to use for Redis")
+        help="the address to use for connecting to a cluster.")
```
Comment (edoakes) [inline python/ray/util/sgd/pytorch/examples/dcgan.py]:
```suggestion
        help="the address to use to connect to a cluster.")
```

**Review by edoakes — COMMENTED** [review]
Looks good! Sorry for the delay on reviewing.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 10]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
 
-Now that the trainer is constructed, you'll naturally want to train the model.
+.. _raysgd-custom-training:
 
-.. code-block:: python
+Custom Training and Validation (Operators)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    trainer.train()
+``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexibility similar to using PyTorch natively.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
``PyTorchTrainer`` allows you to run a custom training and validation loops in parallel on each worker, providing a flexible interface similar to using PyTorch natively.
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
For both training and validation, there are two granularities that you can provide customization - per epoch and per batch. These correspond to ``train_batch``,
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 11]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
+This is done via the :ref:`ref-pytorch-operator` interface.
+
+For both training and validation, there are two granularities that you can provide customization - per epoch or per batch. These correspond to ``train_batch``,
+``train_epoch``, ``validate``, and ``validate_batch``. Other useful methods to override include ``setup``, ``save`` and ``restore``. You can use this class
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
``train_epoch``, ``validate``, and ``validate_batch``. Other useful methods to override include ``setup``, ``save`` and ``restore``. You can use these
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 12]
```diff
@@ -115,30 +118,120 @@ You can also set the number of workers and whether the workers will use GPUs:
...
+to manage state (like a classifier neural network for calculating inception score, or a heavy tokenizer).
 
-This takes one pass over the training data.
+Providing a custom operator is necessary if creator functions return multiple models, optimizers, or schedulers.
 
-To run the model on the validation data passed in by the ``data_creator``, you can simply call:
+Below is a partial example of overriding ``train_batch``.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Below is a partial example of a custom ``TrainingOperator`` that provides a ``train_batch`` implementation for a Deep Convolutional GAN.
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 5 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: fixed.

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 6 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: fixed

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 2 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: fixed.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 13]
```diff
@@ -151,12 +244,11 @@ After training, you may want to reappropriate the Ray cluster. To release Ray re
 Initialization Functions
 ------------------------
 
-You may want to run some initializers on each worker when they are started. This may be something like setting an environment variable or downloading some data. You can do this via the ``initialization_hook`` parameter:
+Use the ``initialization_hook`` parameter to initialize state on each worker process when they are started. This is useful when setting an environment variable or downloading data:
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: done.

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: done

File: python/ray/util/sgd/pytorch/examples/dcgan.py (hunk 8 above)
Comment (richardliaw) [reply python/ray/util/sgd/pytorch/examples/dcgan.py]: Done

**Review by edoakes — APPROVED** [review]
LGTM
